localstack/services/events/provider.py

```diff
@@ -477,6 +514,61 @@ def put_partner_events(
     ) -> PutPartnerEventsResponse:
         raise NotImplementedError
 
+    ######
+    # Tags
+    ######
+
+    @handler("ListTagsForResource")
+    def list_tags_for_resource(
+        self, context: RequestContext, resource_arn: Arn, **kwargs
+    ) -> ListTagsForResourceResponse:
+        store = self.get_store(context)
+        resource_type = get_bus_or_rule(resource_arn)
+        if not resource_type:
+            pass  # TODO handle error for tagging not rule or event_bus
+        self._check_resource_exists(resource_arn, resource_type, store)
+        try:
+            tags = store.TAGS.get(resource_arn, {})
+            tags_list = [{"Key": key, "Value": value} for key, value in tags.items()]
+        except KeyError:
+            raise ResourceNotFoundException(f"Resource not found: '{resource_arn}'")
+        return ListTagsForResourceResponse(Tags=tags_list)
+
+    @handler("TagResource")
+    def tag_resource(
+        self, context: RequestContext, resource_arn: Arn, tags: TagList, **kwargs
+    ) -> TagResourceResponse:
+        # each tag key must be unique
+        # https://docs.aws.amazon.com/general/latest/gr/aws_tagging.html#tag-best-practices
+        store = self.get_store(context)
+        resource_type = get_bus_or_rule(resource_arn)
+        if not resource_type:
+            pass  # TODO handle error for tagging not rule or event_bus
+        self._check_resource_exists(resource_arn, resource_type, store)
+        check_unique_tags(tags)
+        for tag in tags:
+            resource_tags = store.TAGS.setdefault(resource_arn, {})
+            resource_tags.update({tag["Key"]: tag["Value"]} or {})
+
+    @handler("UntagResource")
+    def untag_resource(
+        self, context: RequestContext, resource_arn: Arn, tag_keys: TagKeyList, **kwargs
+    ) -> UntagResourceResponse:
+        store = self.get_store(context)
+        resource_type = get_bus_or_rule(resource_arn)
+        if not resource_type:
+            pass  # TODO handle error for tagging not rule or event_bus
+        self._check_resource_exists(resource_arn, resource_type, store)
+        try:
+            resource_tags = store.TAGS.get(resource_arn, {})
+        except KeyError:
```

Comment:
When can a dict.get() raise a KeyError? This seems untested

Reply:
fixed with using the TaggingService